### DISCUSSION TITLE
Personal Access Token security, expiration and rotation

### DISCUSSION THREAD

**Discussion by mpdude** [community/community#6827] [Other Feature Feedback, Questions, & Ideas]
Recently, the `ua-parser-js` NPM package was compromised (https://github.com/advisories/GHSA-pjwm-rvh2-c87w) to run malicious scripts upon installation. This package is depended-upon by many other NPM packages, also from major organizations and reputable sources. Installing this package as a direct or transitive package dependency puts users at risk.

As a developer installing or updating dozens of packages every day through various package managers, I think it is just a matter of time until you are affected by an issue like that, no matter how cautious or risk-aware you are.

With that assumption, I am particularly worried about GitHub Access tokens that might be found in clear-text in well-known filesystem locations:

* The `gh` CLI keeps an OAuth token in `~/.config/gh/hosts.yml`
* The Composer PHP Package manager [sometimes requires a PAT](https://getcomposer.org/doc/articles/authentication-for-private-packages.md#github-oauth) and keeps it in `~/.composer/auth.json`
* When I use `echo '...PAT...' | docker login ghcr.io -u myuser --password-stdin` (is that _really_ the recommended way of logging into ghcr.io?), the PAT ends up in `~/.docker/config.json`

We use SSH for basic `git` operations because SSH keys seem reasonably secure (passphrase-protected; only on one machine or even on a security token). GitHub.com access [can be configured]((https://docs.github.com/en/organizations/keeping-your-organization-secure/requiring-two-factor-authentication-in-your-organization)) to require 2FA. 

Access tokens seem to be much more vulnerable and have no additional protection.

At least, we could reduce the attack surface/impact if we could scope tokens narrower by default and/or make them expire sooner.

Some ideas on what could be improved:

* `gh` CLI should make it possible to create/use an OAuth token with a limited lifetime only.
* Configure maximum token lifetime per organization-wide setting (not sure if possible, since tokens belong to personal accounts?)
* Provide a command like `aws ecr get-login-password` (through the AWS CLI for the [AWS ECR](https://docs.aws.amazon.com/AmazonECR/latest/userguide/getting-started-cli.html#cli-authenticate-registry)) to simplify/encourage login to `ghcr.io` with short-lived tokens?
* Similar: Make it possible to obtain (short-lived) PATs through the `gh` CLI
* ...?

When tokens expire and re-authentication is necessary (e. g. `gh` CLI needs to re-authenticate every 24h?), this should be user-friendly and simple to do, even if it is running on another machine (SSH session, VM...):

* Maybe showing an URL that you could paste into your browser where you are logged in to GitHub.com?
* Maybe something like OTP/FIDO U2F (don't know the standards and exact specs) on things like a YubiKey?

**Comment by tekumara** [comment]
+1 see also https://github.com/cli/cli/issues/5924
